### PR TITLE
Add better support for application composition.

### PR DESCRIPTION
**Pull request by totherik**
Merely `use`-ing express applications is practical, but doesn't quite go far enough in ensuring application settings are correctly inherited from the parent application. This PR:
- Allows express' built-in prototypical inheritance of settings to continue to be used.
- Changes how kraken handles express application configuration such that existing express settings are preserved and only specified settings are overridden. (Previously kraken replaces express settings.)
- Removes many redundant express setting overrides from default kraken config to allow express to add or remove settings without them having to be tracked here.  (Also allows express to continue to establish defaults).
- Add support for optionally inheriting view configuration (via `options.inheritViews`) from the parent app.
- Supports:
  - kraken apps to be mounted on kraken apps
  - non-kraken apps to be mounted on kraken apps
  - kraken apps to be mounted on non-kraken apps

NOTE: This change _should_ be backward compatible, and once merged to `v1.0.x` can be merged into master.

### CODE REVIEW

**Comment by aredridel** [comment]
`inheritViews` name is a little confusing: It's the configuration, not the views themselves that are inherited, right?

**Comment by totherik** [comment]
nope, the config is always inherited from the parent (unless settings are explicitly overridden by the child app), but a mounted app can specifically opt out of view config inheritance should it want to provide its own rendering (without additional config muckery).

If isolating view config isn't necessary it can certainly be removed, but it seemed that it might be a core use-case.

**Comment by totherik** [comment]
I should add that `express` [doesn't inherit view config by default](https://github.com/strongloop/express/blob/master/lib/application.js#L81-L82) so it seemed like a straightforward way to opt in to that behavior.

**Comment by aredridel** [comment]
Oh that makes sense to call out specifically.

**Comment by aredridel** [comment]
Wow. This feels remarkably solid and do-what-you-want.

File: lib/settings.js [hunk 1]
```diff
@@ -19,22 +19,88 @@
...
 module.exports = function settings(app, options) {
     debug('initializing settings');
 
+    function mount(app, options, config) {
+        var settings, defaults, handler;
+
+        // Get configured settings and express defaults.
+        settings = config.get('express');
+        defaults = Object.keys(app.settings);
+
```
Comment (pvenkatakrishnan) [inline lib/settings.js:36]: doubt: does `app.settings` have only view related settings ?

Comment (aredridel) [inline lib/settings.js:36]: Nope.

File: lib/settings.js [hunk 2]
```diff
@@ -19,22 +19,88 @@
...
+            handler = function (name) {
+                var idx = defaults.indexOf(name);
+                if (idx >= 0) {
+                    defaults.splice(idx, 1);
+                }
+            };
+        }
+
+        // Update our configuration (provided settings or app defaults)
+        ['views', 'view engine', 'view cache'].forEach(handler);
```
Comment (pvenkatakrishnan) [inline lib/settings.js]: yeah so we cherry pick to override or not only view related here.. do we see any other setting that we want to custom configure like this ?

File: lib/settings.js (hunk 1 above)
Comment (totherik) [inline lib/settings.js:36]: In fact, there are [several settings](https://github.com/strongloop/express/blob/656e214937889536b0faa73097422061315496f2/lib/application.js#L53-L59) that [get set](https://github.com/strongloop/express/blob/656e214937889536b0faa73097422061315496f2/lib/application.js#L81-L87), but [never inherited](https://github.com/strongloop/express/blob/656e214937889536b0faa73097422061315496f2/lib/application.js#L68) when an app is mounted.

File: lib/settings.js (hunk 2 above)
Comment (totherik) [inline lib/settings.js]: I mentioned it in an earlier comment, but I expected a typical use-case would be to inherit everything from that parent app save view settings. If there are other settings that should be different, the child app can set them explicitly and they will override the parent settings, but it seemed like view configuration would be the most common use-case of all the settings that might need a switch (so devs don't need to research which property names they need to monkey with to get it working). Definitely open to other suggestions, though.

File: lib/settings.js [hunk 3]
```diff
@@ -19,22 +19,88 @@
...
+    function compose(config) {
+        // If this application is mounted on a parent app we need to make sure
+        // the two work together cleanly and appropriate settings are inherited.
+        if (typeof app.parent === 'function') {
+            // If it has already been mounted do all the work to update settings.
+            mount(app, options, config);
+        } else {
+            // ...if it hasn't been mounted, register a handler for the `mount`
+            // event to apply them as necessary.
+            app.once('mount', function () {
```
Comment (grawk) [inline lib/settings.js:77]: Why aren't you using the `parent` parameter in the onmount callback? http://expressjs.com/4x/api.html#app.onmount

Edit. Nevermind, I see that we need the actual `app` which can be retrieved via `this` in the event listener. Perhaps that is documented somewhere in express but I couldn't find it.

Comment (totherik) [inline lib/settings.js:77]: Node.js sets the `this` context for event listeners; it's not an express thing. And yeah, didn't use parent b/c I didn't need parent. :-)

File: lib/settings.js [hunk 4]
```diff
@@ -19,22 +19,88 @@
...
+
+        if (options.inheritViews) {
+            // If the mounted app SHOULD inherit views, delete from local
+            // kraken settings so the app settings aren't updated later.
+            handler = function (name) {
+                delete settings[name];
+            };
+        } else {
+            // If the views SHOULD NOT be inherited, remove from list
+            // of default settings such that they are NOT deleted delow.
```
Comment (jasisk) [inline lib/settings.js]: `delow`

File: lib/settings.js [hunk 5]
```diff
@@ -19,22 +19,88 @@
...
-        app.set('view', config.get('express:view') ? require(config.get('express:view')) : View);
-        app.set('trust proxy', config.get('express:trust proxy'));
-        app.kraken = config;
+        var settings;
+
+        settings = config.get('express');
+
+        // Allow configuration of custom View impl. This really can be
+        // accomplished using shortstop, but maintain for compatibility.
+        if (typeof settings.view === 'string') {
```
Comment (jasisk) [inline lib/settings.js:93]: Are we planning to deprecate and move—exclusively—to a shortstop handler? If so, we should throw a dep notice.

Comment (totherik) [inline lib/settings.js:93]: I wasn't thinking that we needed to move on that at all yet, or could be in a separate PR. Somewhat orthogonal to this change.

Comment (jasisk) [inline lib/settings.js:93]: Yup, just trying to be more conscious of incremental, informative migratory steps.

**Comment by pvenkatakrishnan** [comment]
:+1:  lgtm

**Comment by grawk** [comment]
I like it. :+1:. But I'll let @jasisk pull the lever since he requested the last change.
